File: apps/api/src/modules/scans/favourites.service.ts
import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import type { FavouriteItem, FavouritesResponse, SharedScanFilters } from '@acme/shared';
import { addFavouriteRequestSchema } from '@acme/shared';
import { ApiError } from '../../shared/errors/api-error';
import { resolveCanonicalProductImageUrl } from '../../shared/utils/product-image';
import { buildProductSearchFilter, normalizeSearchQuery } from '../../shared/utils/product-search';
import { prisma } from '../product-analyze/lib/prisma';
import { buildHistoryAnalysisSummary, matchesSharedScanFilters } from './scan-history-analysis';

const DEFAULT_PAGE_SIZE = 20;
const FILTERED_BATCH_MULTIPLIER = 3;

type FavouriteProduct = {
  id: string;
  barcode: string;
  product_name: string | null;
  brands: string | null;
  image_url: string | null;
  images: unknown;
  nutriscore_grade: string | null;
};

type LatestScanRecord = {
  id: string;
  source: 'barcode' | 'photo';
  overallScore: number | null;
  overallRating: string | null;
  personalAnalysisStatus: 'pending' | 'completed' | 'failed' | null;
  personalResult: unknown;
  multiProfileResult: unknown;
  createdAt: Date;
};

const getValidLimit = (limit?: string): number | undefined => {
  if (!limit) {
    return undefined;
  }

  const parsed = Number.parseInt(limit, 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : undefined;
};

const serializeProduct = (product: FavouriteProduct) => ({
  id: product.id,
  barcode: product.barcode,
  product_name: product.product_name,
  brands: product.brands,
  image_url: resolveCanonicalProductImageUrl(product.image_url, product.images),
  nutriscore_grade: product.nutriscore_grade,
});

const hasSharedScanFilters = (filters: SharedScanFilters): boolean =>
  filters.profileIds.length > 0 || filters.fitBuckets.length > 0;

@Injectable()
export class FavouritesService {
  async getFavourites(
    userId: string,
    cursor?: string,
    limit?: string,
    search?: string,
    filters: SharedScanFilters = { profileIds: [], fitBuckets: [] },
  ): Promise<FavouritesResponse> {
    await this.cleanupOrphanedFavourites(userId);

    const take = getValidLimit(limit) ?? DEFAULT_PAGE_SIZE;
    const normalizedSearch = normalizeSearchQuery(search);
    const where: Prisma.FavoriteWhereInput = {
      userId,
      ...(normalizedSearch
        ? {
            product: {
              is: buildProductSearchFilter(normalizedSearch),
            },
          }
        : {}),
    };

    const [favourites, totalCount] = await Promise.all([
      hasSharedScanFilters(filters)
        ? this.findFilteredFavourites(where, userId, cursor, take + 1, filters)
        : this.findFavourites(where, cursor, take + 1),
      this.countFavourites(where, userId, filters),
    ]);

    const hasMore = favourites.length > take;
    const items = hasMore ? favourites.slice(0, take) : favourites;
    const nextCursor = hasMore ? (items[items.length - 1]?.id ?? null) : null;

    const productIds = items
      .map((favourite) => favourite.product?.id)
      .filter((productId): productId is string => productId != null);
    const latestScansByProductId = await this.findLatestScansForProducts(userId, productIds);

    return {
      items: items.map((favourite) => {
        const scan = favourite.product
          ? latestScansByProductId.get(favourite.product.id)
          : undefined;
        return this.serializeFavouriteItem(favourite, scan);
      }),
      nextCursor,
      totalCount,
    };
  }

  private async countFavourites(
    where: Prisma.FavoriteWhereInput,
    userId: string,
    filters: SharedScanFilters,
  ): Promise<number> {
    if (!hasSharedScanFilters(filters)) {
      return prisma.favorite.count({ where });
    }

    let totalCount = 0;
    let nextCursor: string | undefined;
    const batchSize = 100;

    while (true) {
      const favourites = await this.findFavourites(where, nextCursor, batchSize);

      if (favourites.length === 0) {
        break;
      }

      const productIds = favourites
        .map((favourite) => favourite.product?.id)
        .filter((productId): productId is string => productId != null);
      const latestScansByProductId = await this.findLatestScansForProducts(userId, productIds);

      totalCount += favourites.filter((favourite) => {
        const scan = favourite.product
          ? latestScansByProductId.get(favourite.product.id)
          : undefined;
        const summary = buildHistoryAnalysisSummary(scan?.personalResult, scan?.multiProfileResult);
        return matchesSharedScanFilters(summary, filters);
      }).length;

      if (favourites.length < batchSize) {
        break;
      }

      nextCursor = favourites[favourites.length - 1]?.id;
    }

    return totalCount;
  }

  private async findFavourites(
    where: Prisma.FavoriteWhereInput,
    cursor: string | undefined,
    take: number,
  ) {
    return prisma.favorite.findMany({
      where,
      orderBy: { createdAt: 'desc' },
      take,
      ...(cursor
        ? {
            cursor: { id: cursor },
            skip: 1,
          }
        : {}),
      include: {
        product: {
          select: {
            id: true,
            barcode: true,
            product_name: true,
            brands: true,
            image_url: true,
            images: true,
            nutriscore_grade: true,
          },
        },
      },
    });
  }

  private async findFilteredFavourites(
    where: Prisma.FavoriteWhereInput,
    userId: string,
    cursor: string | undefined,
    take: number,
    filters: SharedScanFilters,
  ) {
    const matches: Array<{
      id: string;
      createdAt: Date;
      product: FavouriteProduct | null;
    }> = [];
    const batchSize = Math.max(take * FILTERED_BATCH_MULTIPLIER, DEFAULT_PAGE_SIZE);
    let nextCursor = cursor;

    while (matches.length < take) {
      const favourites = await this.findFavourites(where, nextCursor, batchSize);

      if (favourites.length === 0) {
        break;
      }

      const productIds = favourites
        .map((favourite) => favourite.product?.id)
        .filter((productId): productId is string => productId != null);
      const latestScansByProductId = await this.findLatestScansForProducts(userId, productIds);

      for (const favourite of favourites) {
        const scan = favourite.product
          ? latestScansByProductId.get(favourite.product.id)
          : undefined;
        const summary = buildHistoryAnalysisSummary(scan?.personalResult, scan?.multiProfileResult);

        if (matchesSharedScanFilters(summary, filters)) {
          matches.push(favourite);
        }

        if (matches.length >= take) {
          break;
        }
      }

      if (favourites.length < batchSize) {
        break;
      }

      nextCursor = favourites[favourites.length - 1]?.id;
    }

    return matches;
  }

  async addFavourite(userId: string, body: unknown): Promise<{ success: true }> {
    const parsed = addFavouriteRequestSchema.safeParse(body);

    if (!parsed.success) {
      throw ApiError.badRequest(parsed.error.issues[0]?.message ?? 'Invalid payload');
    }

    await prisma.favorite.upsert({
      where: {
        userId_productId: {
          userId,
          productId: parsed.data.productId,
        },
      },
      create: {
        userId,
        productId: parsed.data.productId,
      },
      update: {},
    });

    return { success: true };
  }

  async removeFavourite(userId: string, productId: string): Promise<{ success: true }> {
    await prisma.favorite.deleteMany({
      where: { userId, productId },
    });

    return { success: true };
  }

  async getFavouriteStatus(userId: string, productId: string): Promise<{ isFavourite: boolean }> {
    const favourite = await prisma.favorite.findUnique({
      where: { userId_productId: { userId, productId } },
      select: { id: true },
    });

    return { isFavourite: favourite != null };
  }

  private async cleanupOrphanedFavourites(userId: string): Promise<void> {
    const result = await prisma.favorite.deleteMany({
      where: {
        userId,
        product: {
          scans: {
            none: {
              userId,
              type: 'product',
            },
          },
        },
      },
    });

    if (result.count > 0) {
      console.log(`[favourites] Removed ${result.count} orphaned favourite(s) for user=${userId}`);
    }
  }

  private serializeFavouriteItem(
    favourite: {
      id: string;
      createdAt: Date;
      product: FavouriteProduct | null;
    },
    scan?: LatestScanRecord,
  ): FavouriteItem {
    const analysisSummary = buildHistoryAnalysisSummary(
      scan?.personalResult,
      scan?.multiProfileResult,
    );

    return {
      favouriteId: favourite.id,
      id: scan?.id ?? favourite.id,
      type: 'product',
      createdAt: (scan?.createdAt ?? favourite.createdAt).toISOString(),
      source: scan?.source ?? 'barcode',
      overallScore: scan?.overallScore ?? null,
      overallRating: scan?.overallRating ?? null,
      personalScore: analysisSummary.personalScore,
      personalRating: analysisSummary.personalRating,
      personalAnalysisStatus: scan?.personalAnalysisStatus ?? null,
      mainUserHasDietConflict: analysisSummary.mainUserHasDietConflict,
      isFavourite: true,
      profileChips: analysisSummary.profileChips,
      product: favourite.product ? serializeProduct(favourite.product) : null,
    };
  }

  private async findLatestScansForProducts(
    userId: string,
    productIds: string[],
  ): Promise<Map<string, LatestScanRecord>> {
    if (productIds.length === 0) {
      return new Map();
    }

    const scans = await prisma.scan.findMany({
      where: {
        userId,
        type: 'product',
        productId: { in: productIds },
      },
      orderBy: { createdAt: 'desc' },
      distinct: ['productId'],
      select: {
        id: true,
        productId: true,
        source: true,
        overallScore: true,
        overallRating: true,
        personalAnalysisStatus: true,
        personalResult: true,
        multiProfileResult: true,
        createdAt: true,
      },
    });

    const scansByProductId = new Map<string, LatestScanRecord>();
    for (const scan of scans) {
      if (scan.productId) {
        scansByProductId.set(scan.productId, {
          id: scan.id,
          source: scan.source,
          overallScore: scan.overallScore,
          overallRating: scan.overallRating,
          personalAnalysisStatus: scan.personalAnalysisStatus,
          personalResult: scan.personalResult,
          multiProfileResult: scan.multiProfileResult,
          createdAt: scan.createdAt,
        });
      }
    }

    return scansByProductId;
  }
}

File: apps/server-py/app/services/favourites_service.py
from typing import Optional

from sqlalchemy import and_, or_, select
from sqlalchemy.orm import selectinload

from app.models.favorite import Favorite
from app.models.scan import Scan
from app.enums import ScanType
from app.schemas.favourites import (
    FavouriteItem,
    FavouriteListResponse,
)
from app.schemas.scans import ProductSummary
from app.services.scans_service import _extract_profile_chips
from app.utils.unitofwork import UnitOfWork

DEFAULT_PAGE_SIZE = 20


class FavouritesService:
    async def get_list(
        self,
        uow: UnitOfWork,
        *,
        user_id: str,
        cursor: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> FavouriteListResponse:
        take = limit or DEFAULT_PAGE_SIZE

        async with uow:
            stmt = (
                select(Favorite)
                .where(Favorite.user_id == user_id)
                .options(selectinload(Favorite.product))
                .order_by(Favorite.created_at.desc(), Favorite.id.desc())
                .limit(take + 1)
            )

            if cursor:
                cursor_stmt = select(Favorite.created_at, Favorite.id).where(Favorite.id == cursor)
                cursor_row = (await uow.session.execute(cursor_stmt)).one_or_none()
                if cursor_row:
                    stmt = stmt.where(
                        or_(
                            Favorite.created_at < cursor_row.created_at,
                            and_(
                                Favorite.created_at == cursor_row.created_at,
                                Favorite.id < cursor,
                            ),
                        )
                    )

            result = await uow.session.execute(stmt)
            favourites = list(result.scalars().all())

            # Get latest scan per product for these favourites
            product_ids = [f.product.id for f in favourites if f.product]
            scan_map: dict[str, Scan] = {}
            if product_ids:
                # Subquery: latest scan id per product_id for this user
                from sqlalchemy import func

                sub = (
                    select(
                        Scan.product_id,
                        func.max(Scan.created_at).label("max_created_at"),
                    )
                    .where(
                        Scan.user_id == user_id,
                        Scan.type == ScanType.product,
                        Scan.product_id.in_(product_ids),
                    )
                    .group_by(Scan.product_id)
                    .subquery()
                )

                scans_stmt = (
                    select(Scan)
                    .join(
                        sub,
                        and_(
                            Scan.product_id == sub.c.product_id,
                            Scan.created_at == sub.c.max_created_at,
                        ),
                    )
                    .where(Scan.user_id == user_id, Scan.type == ScanType.product)
                )

                scans_result = await uow.session.execute(scans_stmt)
                for scan in scans_result.scalars().all():
                    if scan.product_id and scan.product_id not in scan_map:
                        scan_map[scan.product_id] = scan

        has_more = len(favourites) > take
        items_raw = favourites[:take]
        next_cursor = items_raw[-1].id if has_more and items_raw else None

        items = []
        for fav in items_raw:
            scan = scan_map.get(fav.product_id) if fav.product_id else None
            raw = (scan.personal_result or {}) if scan else {}
            profiles = raw.get("profiles", []) if isinstance(raw, dict) else []
            first = profiles[0] if profiles else {}
            personal_score = first.get("score") if isinstance(first, dict) else None
            personal_rating = first.get("fitLabel") if isinstance(first, dict) else None

            product_summary = None
            if fav.product:
                product_summary = ProductSummary(
                    id=fav.product.id,
                    barcode=fav.product.barcode,
                    product_name=fav.product.product_name,
                    brands=fav.product.brands,
                    image_url=fav.product.image_url,
                    nutriscore_grade=fav.product.nutriscore_grade,
                )

            items.append(
                FavouriteItem(
                    favouriteId=fav.id,
                    id=scan.id if scan else fav.id,
                    type="product",
                    createdAt=(scan.created_at if scan else fav.created_at).isoformat(),
                    source=scan.source if scan else "barcode",
                    overallScore=scan.overall_score if scan else None,
                    overallRating=scan.overall_rating if scan else None,
                    personalScore=personal_score,
                    personalRating=personal_rating,
                    personalAnalysisStatus=scan.personal_analysis_status if scan else None,
                    isFavourite=True,
                    profileChips=_extract_profile_chips(scan.multi_profile_result if scan else None),
                    product=product_summary,
                )
            )

        return FavouriteListResponse(items=items, nextCursor=next_cursor)

    async def add(self, uow: UnitOfWork, *, user_id: str, product_id: str) -> None:
        async with uow:
            existing = await uow.favorites.get_one_or_none(user_id=user_id, product_id=product_id)
            if existing is None:
                await uow.favorites.create({"user_id": user_id, "product_id": product_id})

    async def remove(self, uow: UnitOfWork, *, user_id: str, product_id: str) -> None:
        async with uow:
            await uow.favorites.delete(return_object=False, user_id=user_id, product_id=product_id)

    async def is_favourite(self, uow: UnitOfWork, *, user_id: str, product_id: str) -> bool:
        async with uow:
            existing = await uow.favorites.get_one_or_none(user_id=user_id, product_id=product_id)
        return existing is not None
